sdk/azidentity/workload_identity.go

```diff
@@ -0,0 +1,78 @@
+//go:build go1.18
+// +build go1.18
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package azidentity
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+)
+
+// WorkloadIdentityCredential supports Azure Workload Identity on Kubernetes.
+// See [AKS documentation] for more information.
+//
+// [AKS Workload Identity documentation]: https://learn.microsoft.com/azure/aks/workload-identity-overview
+type WorkloadIdentityCredential struct {
+	assertion, file string
+	cred            *ClientAssertionCredential
+	expires         time.Time
+	mtx             *sync.RWMutex
+}
+
+// WorkloadIdentityCredentialOptions contains optional parameters for WorkloadIdentityCredential.
+type WorkloadIdentityCredentialOptions struct {
+	azcore.ClientOptions
+}
+
+// NewWorkloadIdentityCredential constructs a WorkloadIdentityCredential. tenantID and clientID specify the identity the credential authenticates.
+// file is a path to a file containing an assertion that authenticates the identity.
+func NewWorkloadIdentityCredential(tenantID, clientID, file string, options *WorkloadIdentityCredentialOptions) (*WorkloadIdentityCredential, error) {
+	if options == nil {
+		options = &WorkloadIdentityCredentialOptions{}
+	}
+	w := WorkloadIdentityCredential{file: file, mtx: &sync.RWMutex{}}
+	cred, err := NewClientAssertionCredential(tenantID, clientID, w.getAssertion, &ClientAssertionCredentialOptions{ClientOptions: options.ClientOptions})
+	if err != nil {
+		return nil, err
+	}
+	cred.name = fmt.Sprintf("%T", w)
+	w.cred = cred
+	return &w, nil
+}
+
+// GetToken requests an access token from Azure Active Directory. Azure SDK clients call this method automatically.
+func (w *WorkloadIdentityCredential) GetToken(ctx context.Context, opts policy.TokenRequestOptions) (azcore.AccessToken, error) {
+	return w.cred.GetToken(ctx, opts)
+}
+
+// getAssertion returns the specified file's content, which must be an appropriately formatted assertion (the platform is responsible for formatting)
+func (w *WorkloadIdentityCredential) getAssertion(context.Context) (string, error) {
+	w.mtx.RLock()
```

Comment:
Would using the `ExpiringResource` from the `internal` module be advantageous here?

Reply:
I don't think so. `ExpiringResource` implements a fault-tolerant refresh window during which it attempts to update the resource before it expires. In this case we just want to ensure we read the file at most once every 5 minutes. The platform is responsible for keeping the file up to date, and we can't establish a refresh window around it because we don't know when the content expires.

Reply:
Thanks for the explanation.  Might be good to add a comment about the platform keeping the file current, and/or a link to any docs that explains how this works.

Reply:
You mean, replace this lock upgrade/double check approach with waiting on a `sync.Cond`?

Reply:
Correct.

Reply:
Here's what I came up with: [gist](https://gist.github.com/chlowell/3a7b9491f6c0b0aaf0c79695b0d3103c). I find it harder to read (and to write!). I also found it's marginally slower than the `RWMutex` approach, I guess because goroutines must acquire an exclusive lock before calling `Cond.Wait()`, and that overhead doesn't pay off because we're waiting on relatively fast I/O. Am I overlooking some other possible benefit in using `sync.Cond`?

Reply:
My thought process was, if we have an algorithm for concurrently checking/acquiring/updating shared state (OAuth tokens), then the same algorithm should apply here too.  But perhaps this case is different enough (local file IO, no refresh window) that it doesn't matter as much.

Maybe @JeffreyRichter has an opinion on this.

Reply:
I'll also add that what you have is correct and can always be changed later, so we don't have to block on this now if we want to get the feature out.